api/v1alpha1/edgeplaybook_types.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// EdgePlaybookSpec defines the desired state of EdgePlaybook
+type EdgePlaybookSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	DeviceSelector *metav1.LabelSelector `json:"deviceSelector,omitempty"`
+
+	// The ansible playbook command to execute
+	AnsiblePlaybookCmd *AnsiblePlaybookCmd `json:"ansiblePlaybookCmd,omitempty"`
+
+	LogCollection map[string]*LogCollectionConfig `json:"logCollection,omitempty"`
+}
+
+type AnsiblePlaybookCmd struct {
+	// username who execute the playbook
+	User string `json:"user,omitempty"`
+	// Ansible binary file
+	BinaryPath string `json:"binaryPath,omitempty"`
+	// the ansible's playbooks list to be used
+	Playbooks []Playbook `json:"playbooks,omitempty"`
+	// the ansible's playbook options
+	Options *AnsibleOptions `json:"ansibleOptions,omitempty"`
+	// the ansible's playbook privilege escalation options
+	PrivilegeEscalationOptions *PrivilegeEscalationOptions `json:"privilegeEscalationOptions,omitempty"`
+	// enum that defines which is the stdout callback method (default = 'default'). Supported stdout methods are: debug,default,dense,json,minimal,null,oneline,stderr,timer,yaml,runner-events
+	StdoutCallback string `json:"stdoutCallback,omitempty"`
+}
+
+type AnsibleOptions struct {
+	// don't make any changes; instead, try to predict some of the changes that may occur
+	Check bool `json:"check,omitempty"`
+}
+
+type PrivilegeEscalationOptions struct {
+	Become bool `json:"become,omitempty"`
+	// Enum ksu, pbrun, enable, sesu, pmrun,runas, sudo, su, doas, pfexec, Machinectl, dzdo
+	BecomeMethod string `json:"becomeMethod,omitempty"`
+	BecomeUser   string `json:"becomeUser,omitempty"`
+}
+
+type Playbook struct {
+	// Playbook content
+	Content string `json:"content"`
```

Comment:
I think this is not enough. Playbook could be simple one file, but could be also complex directory structure.

Reply:
I think we could use following:
1) web link to a content I think archived, liked tar.gz file
2) Use the S3 as we do for workloads, but the other way around.
3) Use Ansible Tower/AWX links, I am not sure if they have such functionality, but maybe worth to explore.

I think for start we can use web link.

Reply:
Can we, for simplicity, start with in-line playbook and expand with options proposed by @machacekondra  later?

Reply:
I replaced this with a URL string (easy to undo) but in my opinion @jakub-dzon is right. We can make this CRD more complex later

Reply:
I think you can make the field '[]byte'. Think of it as an HTTP body. it can be anything. So it can be a single text file. It can be an archive of any type. this way in the future u can easily support the archives that the web link would have fetched. There are code libraries that will tell u the type of the content so u don't even need a separate field.

Reply:
Thanks @ydayagi for your suggestion! It make sense to me. 

Can you please give me an example of
>  "There are code libraries that will tell u the type of the content so u don't even need a separate field."

I've probably never used them :) 

@jakub-dzon @machacekondra do you agree with this?

Reply:
well...the 'file' command in linux uses libmagic (or something like that). I'm sure there are libraries in GO. But even if u add a field for specifying the type it does not make a great difference. I googled and got this one: https://github.com/vimeo/go-magic

Reply:
seems u can even use http.DetectContentType(bytes)

Reply:
Done